File: fair_extension/assets/plugin/fair_net_plugin.js
let FairNetCallBack = {};
let callBackId = 0;

class FairNet {

	static GET = 'GET';
	static POST = 'POST';

	static requestData(req) {
		let respMap = {};
		let id = 'FairNet$' + (callBackId++);
		// 设置回调
		let reqFunc = {};
		if (req.complete) {
			reqFunc['complete'] = req.complete;
		}
		if (req.success) {
			reqFunc['success'] = req.success;
		}
		if (req.failure) {
			reqFunc['failure'] = req.failure;
		}
		FairNetCallBack[id] = reqFunc;
		// 处理参数
		let method = '';
		if (req.method) {
			method = req.method;
		}
		let url = '';
		if (req.url) {
			url = req.url;
		}
		let data = {};
		if (req.data) {
		    data = mapOrSetToObject(req.data);
		}
		let reqMap = {
			pageName: '#FairKey#',
			funcName: 'invokePlugin',
			'className': 'FairNet#requestData',
			args: {
				callId: id,
				method: method,
				url: url,
				data: data
			}
		};
		invokeFlutterCommonChannel(JSON.stringify(reqMap), (resultStr) =>{
			let responseMap = JSON.parse(resultStr);
			let data = responseMap['data'];
			responseMap['data'] = data.data;
			let id = responseMap['callId'];
			//处理需要返回的结果值
			let callback = FairNetCallBack[id];
			if (callback == null) {
				return
			}
			let complete = callback['complete'];
			let failure = callback['failure'];
			let success = callback['success'];
			if (responseMap['statusCode'] === 200) {
				if (complete != null) {
					complete();
				}
				if (success != null) {
					success(convertObjectLiteralToSetOrMap(responseMap));
				}
			} else {
				if (complete != null) {
					complete();
				}
				if (failure != null) {
					failure(responseMap['statusMessage']);
				}
			}
		})
	}
}

File: fair_extension/lib/net/fair_net_plugin.dart
import 'dart:convert';

import 'package:dio/dio.dart';
import 'package:fair/fair.dart';

/// 网络请求
class FairNet extends IFairPlugin {
  static final FairNet _fairNet = FairNet._internal();

  static const GET = 'GET';
  static const POST = 'POST';

  FairNet._internal();

  factory FairNet() {
    return _fairNet;
  }

  static Future<dynamic> requestData(
      {required String method,
      required String url,
      Map? data,
      void Function()? complete,
      void Function(Map resp)? success,
      void Function(String msg)? failure}) async {
    // 空实现，仅用dart端占位（实现见fair_net_plugin.js）
  }

  static Future<dynamic> _request(dynamic map) async {
    if (map == null) {
      return;
    }
    var req;
    bool isDart;
    if (map is Map) {
      isDart = true;
      req = map;
    } else {
      isDart = false;
      req = jsonDecode(map);
    }
    var pageName = req['pageName'];
    var args = req['args'];
    if (isDart) {
      args = req;
    }

    var url = args['url'];
    var callId = args['callId'];
    var successCallback = args['success'];
    var failureCallback = args['failure'];
    var completeCallback = args['complete'];
    Map<String, dynamic> reqData = args['data'];
    var method = args['method'];
    Response<dynamic>? response;

    if (method == null) {
      return Future.value();
    }

    switch (method) {
      case 'GET':
        response = await _get(url, queryParameters: reqData);
        break;
      case 'POST':
        response = await _post(url, queryParameters: reqData);
        break;
    }

    var statusCode = response?.statusCode;
    var data = response?.data;
    var statusMessage = response?.statusMessage;

    /// 需要判断发起方的请求是dart端还是js端
    if (isDart) {
      /// 实际处理结合自身app的业务逻辑场景
      if (200 == statusCode) {
        successCallback?.call(data);
        completeCallback?.call();
      } else {
        failureCallback?.call(statusMessage);
        completeCallback?.call();
      }
      return Future.value();
    } else {
      var resp = {
        'callId': callId,
        'pageName': pageName,
        'statusCode': response?.statusCode,
        'data': response?.data,
        'statusMessage': response?.statusMessage,
      };
      return Future.value(jsonEncode(resp));
    }
  }

  static Future<Response<dynamic>> _post(String path,
      {Map<String, dynamic>? queryParameters}) async {
    var resp = await _getDio().post(path, queryParameters: queryParameters);
    return Future.value(resp);
  }

  static Future<Response<dynamic>> _get(String path,
      {Map<String, dynamic>? queryParameters}) async {
    var resp = await _getDio().get(path, queryParameters: queryParameters);
    return Future.value(resp);
  }

  @override
  Map<String, Function> getRegisterMethods() {
    var functions = <String, Function>{};
    functions.putIfAbsent('requestData', () => _request);
    return functions;
  }

  static Dio? _dio;

  static Dio _getDio() {
    _dio ??= Dio();
    return _dio!;
  }
}
